File: java/src/n03_animal_race/Animal.java
package n03_animal_race;

import java.util.Random;

public abstract class Animal {
    public final String name;
    private final float minSpeed;
    private final float maxSpeed;
    private final int stamina;

    public Animal(String name, float minSpeed, float maxSpeed, int stamina) {
        this.name = name;
        this.minSpeed = minSpeed;
        this.maxSpeed = maxSpeed;
        this.stamina = stamina;
    }

    public float move() {
        return new Random().nextFloat(this.minSpeed, this.maxSpeed)
                * (10f / this.stamina);
    }

    public abstract void makeNoise();
}

File: dart/src/n03_animal_race/animal.dart
import 'dart:math';

abstract class Animal {
  Animal(this.name, this._minSpeed, this._maxSpeed, this._stamina);

  final String name;
  final double _minSpeed;
  final double _maxSpeed;
  final int _stamina;

  double move() => (Random().nextDouble()
      * (this._maxSpeed - this._minSpeed) + this._minSpeed)
      * (10 / this._stamina);

  void makeNoise();
}
